prov/shm: add large message segmentation fallback protocol

aingerson opened the pull request.

aingerson commented:
@shefty @zhngaj @rajachan 
This is a rough draft - I'm still looking for a bug with the SAR write protocol.
As I mentioned, the best implementation for this requires a complete overhaul of the shm addressing. Those changes are also a work in progress but I don't think they are going to be ready for 1.10.
The issues are regarding making sure peers are mapped when transfers need a response entry and storing/tracking per-peer information properly to avoid one peer consuming the entire cmd queue. The provider currently has no way to prevent a peer from consuming the entire queue which is a problem but not something that will necessarily break everything, just cause performance issues. However, for SAR transfers, this is needed to maintain ordering. This patch set gets around that adding a state[MAX_PEERS] per EP to make sure we only send one SAR message to a peer at a time. This is not ideal but solves the issue and eventually would be removed when the addressing/per-peer data refactoring happens.

rwespetal commented:
bot:aws:retest

aingerson commented:
@zhngaj Updated with the fix for the write error I was getting. Should all be working now. I've tested it by forcing CMA off and increasing the sar_threshold pretty high. I haven't encountered any errors. I've also run it through the ubertest using only the SAR protocol and everything seems to be working. Could you do some testing with these changes and let me know what you see?

zhngaj commented:
> @zhngaj Updated with the fix for the write error I was getting. Should all be working now. I've tested it by forcing CMA off and increasing the sar_threshold pretty high. I haven't encountered any errors. I've also run it through the ubertest using only the SAR protocol and everything seems to be working. Could you do some testing with these changes and let me know what you see?

Thank you, Alexia. Really glad to know it's working.
Our nightly CI will test against it extensively. I will also do some manual testing in the meantime.

aingerson commented:
@zhngaj I can't seem to reproduce the EFA failures. They all seem to be efa/rxr errors:
```
prov/efa/src/rxr/rxr_pkt_cmd.c:492: rxr_pkt_handle_recv_completion: Assertion `pkt_entry->pkt_size > 0' failed.
```

Could you help me understand the cause of these failures? Something to do with improper completion reporting maybe?

wzamazon commented:
@aingerson  the failure
```
prov/efa/src/rxr/rxr_pkt_cmd.c:492: rxr_pkt_handle_recv_completion: Assertion `pkt_entry->pkt_size > 0' failed.
```
means we received a packet without any content.

```pkt_entry->pkt_size``` was set as ```cq_entry->len```. It being 0, would indicate ```cq_entry->len``` is 0.

zhngaj commented:
> can't seem to reproduce the EFA failures. They all seem to be efa/rxr errors:

Which test were you running into this error?

aingerson commented on include/ofi_shm.h:
```diff
@@ -227,9 +230,24 @@ struct smr_inject_buf {
+struct smr_sar_buf {
+	uint64_t	status;
+	uint8_t		buf[SMR_SAR_SIZE];
+};
```
What do you think is the best way to store this status? Have the SAR_SIZE be smaller to align the whole buffer?

shefty replied:
I don't have a great answer here.  Unless buf is aligned and a multiple of page size, copy routines are going to get out of whack with either the src or dst buffer.  So, I wouldn't make any changes to this now.

aingerson replied:
I can also include this in the peer data once I refactor the addressing/per peer data. Since a peer only uses one SAR buffer at a time, we could just look up the SAR status for that peer.

shefty replied:
Btw, the buffer pool will throw off all alignment unless it is specified as part of the pool attributes.  So, we can ignore this for quite a while I think.

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -570,9 +570,17 @@ static int smr_progress_cmd_rma(struct smr_ep *ep, struct smr_cmd *cmd)
+			resp->status = -err;
+		} else {
+			ep->region->cmd_cnt++;
+		}
```
Probably, I'll take a look

aingerson commented on prov/shm/src/smr.h:
```diff
@@ -201,8 +217,11 @@ struct smr_ep {
 	struct smr_queue	unexp_msg_queue;
 	struct smr_queue	unexp_tagged_queue;
+	struct dlist_entry	sar_list;
+	enum smr_peer_state	state[SMR_MAX_PEERS];
```
Ideally it would be. The problem is that the smr_ep doesn't currently hold any per peer information. Either it's in the av or it's shared in the EP memory region. I'm fine moving this to the shared memory region. The main reason I didn't is because it doesn't need to be shared by the peers and the memory is called "peer_addr" which this isn't really. But I'm fine either renaming or looking the other way.
But this is also the general source of my comment earlier that to properly implement this (and to fix the shm provider) we would need to refactor the entire addressing/peer exchange structure to have all the correct peer data (including cmd_cnt, sar_cnt) in the shared memory per-peer data. The reason we can't truly do it right now (which I realized while implementing this) is that the target peer is not always guaranteed to have the initiating peer inserted/mapped which means none of this peer_addr data is going to be valid and also means transfers that require a response entry (iov, mmap, sar, or DELIVERY_COMPLETE) will fail if the peer has not inserted the initiator into the AV yet.
I'm working on fixing this by making sure the peer maps the shared memory with a RTS/CTS type of protocol (sort of like how rxd exchanges the dg addresses) 
I know that was a lot of words... feel free to message me for alternate words.

shefty replied:
okee-dokee

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -308,6 +352,52 @@ static int smr_progress_mmap(struct smr_cmd *cmd, struct iovec *iov,
+
+	sar_msg = (struct smr_sar_msg *) smr_get_addr(ep->region, cmd->msg.data.sar);
+	peer_smr = smr_peer_region(ep->region, cmd->msg.hdr.addr);
+	resp = (struct smr_resp *) smr_get_addr(peer_smr, cmd->msg.hdr.src_data);
```
The function essentially is pretty much the equivalent to just adding the parameters. Basically just given the offset and the start of the smr, returns the address of that offset. The callers cast that void * address into the type they need/know that offset is of. Do we have to cast it from void *? Will gcc complain without the cast?

shefty replied:
I would rename smr_get_addr to smr_get_ptr, and convert it from a macro to a static inline that returns a void *.  That can be done separately and later.  Addr makes me think of a network address, not a virtual memory address.

aingerson replied:
okee-dokee

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -726,6 +832,56 @@ static void smr_progress_cmd(struct smr_ep *ep)
+			dlist_remove(&sar_entry->entry);
+			freestack_push(ep->sar_fs, sar_entry);
+		}
+
```
😈

zhngaj commented on prov/util/src/util_shm.c:
```diff
@@ -97,12 +98,18 @@ int smr_create(const struct fi_provider *prov, struct smr_map *map,
+
+	//TODO - find better way to calculate number of SAR buffers
+	//(at least 1 but can't be outrageous)
+	sar_cnt = MAX(1, rx_size / 4);
```
Is there any reason we user rx_size/4 here? IIRC, the rx_size is hard-coded to 1024, then why we need to compare with 1?

aingerson replied:
This calculation and the comment above are outdated from earlier changes when peers could have more than 1 pending SAR transfer at a time.
Since the maximum number of peers is hard coded to 256 right now and every peer is only allowed to have one outstanding SAR transfer at a time, we will only ever need SMR_MAX_PEERS number of SAR buffers. We can change it to SMR_MAX_PEERS instead. Does that seem reasonable to you?

zhngaj replied:
Yup, sounds reasonable to me. Please add the comment when you change it, so users can get a little context.

zhngaj commented on include/ofi_shm.h:
```diff
@@ -156,6 +156,7 @@ struct smr_cmd {
 
 #define SMR_INJECT_SIZE		4096
 #define SMR_COMP_INJECT_SIZE	(SMR_INJECT_SIZE / 2)
+#define SMR_SAR_SIZE		(1 << 14)	//16k
```
It's better to unify here, either use 16384 or change SMR_INJECT_SIZE to (1 << 12)

zhngaj commented on prov/shm/src/smr_init.c:
```diff
@@ -37,7 +37,14 @@
 
 extern struct sigaction *old_action;
+struct smr_env smr_env = {
+	.sar_threshold = SMR_INJECT_SIZE * 8, //32k
```
SMR_SAR_SIZE * 2

zhngaj commented:
@aingerson, given the performance results you posted and the timeline for 1.10, I think it's better and safer for us to fallback to your segmentation protocol and skip mmap protocol for now. And we can re-visit the mmap performance after 1.10 or later. If it sounds reasonable to you, could you please have a separate patch to skip mmap (maybe set sar_thresold to MAX_SIZE?)

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -85,19 +124,24 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_tx_entry *pendi
+		smr_freestack_push(smr_sar_pool(peer_smr), sar_msg);
+		peer_smr->sar_cnt++;
+		ep->state[pending->addr] = SMR_PEER_READY;
+	}
```
I organized it this way to minimize the locked code. We only need to lock the region when pushing resources back in. If we lock while progressing the entry, we lose getting to progress the first SAR buffer while the peer copies the second buffer. But then we also risk having to reprogress/recheck these resources when we can actually get the lock. Thoughts?

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -726,6 +832,56 @@ static void smr_progress_cmd(struct smr_ep *ep)
+		else
+			smr_try_progress_from_sar(sar_msg, resp, &sar_entry->cmd,
+					sar_entry->iov, sar_entry->iov_count,
+					&sar_entry->bytes_done);
```
The problem is that the direction/op check switches depending on whether we're trying to progress from the sender or receiver. Ie the sender wants to progress into the sar if we're doing a write but out if we're doing a read and the receiver wants to do the opposite (out of the sar on a write and in on a read). I'll think if I can simplify these checks more.

aingerson commented:
Updated to address comments and add some refactoring/fixes

wzamazon commented:
@aingerson  The EFA failure
```
fi_rma_bw: prov/util/src/util_buf.c:220: ofi_bufpool_destroy: Assertion `(pool->attr.flags & OFI_BUFPOOL_NO_TRACK) || (buf_region->use_cnt == 0)' failed.
bash: line 1: 63750 Aborted (core dumped) /home/ubuntu/libfabric/fabtests/install/bin/fi_rma_bw -e rdm -o writedata -p "efa" -E 127.0.0.1
```

Was due to a EFA bug, and will be fixed by https://github.com/ofiwg/libfabric/pull/5818

BTW, when I was testing an earlier version of this PR, I encountered data validation error when running fi_bw test with EFA. It is not showing up with the newly updated PR, I suppose it is because you fixed something?

aingerson commented:
@wzamazon Oh awesome, thank you. I was just in the middle of asking you about it.
Yes, the validation error I fixed. I was trying to figure out how the race condition would happen and couldn't map it out, but implemented a way to make sure to avoid it and that seems to have been the issue. It was that the receiver was copying the buffers out of order. I fixed this by adding a field to track which buffer to copy next to make sure we get the correct data.
I'm making sure the FI_MULTI_RECV completion logic is correct with the new protocol. There might be one more bug and I will repush with a fix for that if needed or confirm that there are no issues.

wzamazon commented:
@aingerson https://github.com/ofiwg/libfabric/pull/5818 was merged, if you rebase this PR to master, I believe AWS CI would pass.

aingerson commented:
Updated with MULTI_RECV fix and rebased to pass AWS CI

aingerson commented:
@wzamazon @zhngaj Looks like the AWS error is unrelated. I think it passed the single node tests but let me know if I'm wrong. Once you do any extra testing you want with this PR, update me on any extra changes you would like from these patches.

@shefty I think I addressed all your comments. Let me know if you have more.

wzamazon commented:
@aingerson That nccl-multi-node failure is a transient instance issue. I will retest just to be sure.

wzamazon commented:
bot:aws:retest
